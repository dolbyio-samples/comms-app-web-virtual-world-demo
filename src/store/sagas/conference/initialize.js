import { select, put, call } from "redux-saga/effects";

import { application } from "../../effects";
import { initializeSDK } from "../../services/sdk";
import { retrieveAccessToken } from "../../services/authentication";

export const getToken = (state) => state.application.authentication.token;

export default function* initialize() {
  const clientAccessToken = import.meta.env.VITE_CLIENT_ACCESS_TOKEN;

  // use access token stored in env file. Note: refresh token won't do anything here
  if (clientAccessToken) {
    yield call(initializeSDK, {
      token: clientAccessToken,
      callback: () => {},
    });
  } else {
    // get client access token from your own authentication server.
    // get helper function to get a token
    const refreshToken = yield call(retrieveAccessToken);

    // get token
    const token = yield select(getToken);

    yield call(initializeSDK, {
      token,
      callback: refreshToken,
    });
  }

  yield put(application.sdkInitialized());
}
