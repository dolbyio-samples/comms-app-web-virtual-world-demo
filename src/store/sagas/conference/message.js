import { createId } from "@paralleldrive/cuid2";
import { addMessage } from "../../effects/message";
import { application } from "../../effects";
import { put, call } from "redux-saga/effects";

export function* displayMessage({ type, message }) {
  const id = yield call(createId);
  yield put(
    addMessage({
      entityType: "messages",
      data: {
        [id]: {
          type,
          message,
        },
      },
    })
  );

  yield put(application.addMessageID({ id }));
}
