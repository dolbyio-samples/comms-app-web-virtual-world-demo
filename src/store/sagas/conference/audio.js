import { call } from "redux-saga/effects";

import { getEntity } from "../../reducers/entities/selectors";
import {
  startLocalAudio as start,
  stopLocalAudio as stop,
} from "../../services/audio";
import { displayMessage } from "./message";

// FIXME: create a selectors files ?
export const getParticipant = (state, id) =>
  getEntity(state, "participants", id);

export function* startAudio() {
  const { error } = yield call(start);

  if (error) {
    yield call(displayMessage, {
      type: "error",
      message: "could not start audio",
    });
  }
}
export function* stopAudio() {
  const { error } = yield call(stop);

  if (error) {
    yield call(displayMessage, {
      type: "error",
      message: "could not stop audio",
    });
  }
}
