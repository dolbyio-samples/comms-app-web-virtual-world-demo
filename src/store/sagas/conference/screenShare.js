import { call } from "redux-saga/effects";

import {
  startScreenShare as start,
  stopScreenShare as stop,
} from "../../services/screenShare";
import { displayMessage } from "./message";

export function* startScreenShare(action) {
  const { error } = yield call(start, { audio: true });
  if (error) {
    yield call(displayMessage, {
      type: "error",
      message: "could not start screen sharing",
    });
  }
}

export function* stopScreenShare() {
  const { error } = yield call(stop);
  if (error) {
    yield call(displayMessage, {
      type: "error",
      message: "could not stop screen sharing",
    });
  }
}
