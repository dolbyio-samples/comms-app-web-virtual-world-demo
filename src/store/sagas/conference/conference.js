import { call, put, select } from "redux-saga/effects";

import { application } from "../../effects";
import { push } from "redux-first-history";

import {
  createConference,
  getLocalParticipant,
  joinConference,
  leaveConference,
} from "../../services/conference";
import { getCurrentConferenceID } from "../../reducers/application/selectors";
import { getEntity } from "../../reducers/entities/selectors";
import {
  addConference,
  addParticipant,
  addStreamToMainStage,
  addStreamToSecondStage,
  removeStreamToMainStage,
  removeStreamToSecondStage,
  updateConference,
} from "../../effects/conference";

import { getParticipant } from "./audio";

export const hasVideo = (streams) => {
  const stream = streams.find((stream) => stream.type === "Camera");

  if (!stream) return false;

  const tracks = stream.getVideoTracks();
  return tracks.length > 0;
};

export function* startBroadcastVideoOnMainStage({ payload }) {
  const state = yield select();

  const conferenceID = yield select(getCurrentConferenceID);

  //1- get participant
  const participant = yield select(getParticipant, payload.participantId);

  if (participant) {
    const streams = participant.streams;

    if (streams && streams.length > 0) {
      const stream = yield call(
        getEntity,
        state,
        "streams",
        participant.streams[0]
      );

      //2- check the stream has video track
      if (stream) {
        const video = yield call(hasVideo, [stream]);

        //3- if so add stream id to stream ids array to be displayed (update conference entities)
        if (video) {
          yield put(
            addStreamToMainStage({
              id: conferenceID,
              entityType: "conferences",
              data: {
                mainStageStreams: [participant.streams[0]],
              },
            })
          );
        }
      }
    }
  }
}

export function* stopBroadcastVideoOnMainStage({ payload }) {
  const conferenceID = yield select(getCurrentConferenceID);

  yield put(
    removeStreamToMainStage({
      id: conferenceID,
      entityType: "conferences",
      data: {
        mainStageStreams: [],
      },
    })
  );
}

export function* startBroadcastVideoOnSecondStage({ payload }) {
  const state = yield select();

  const conferenceID = yield select(getCurrentConferenceID);

  //1- get participant
  const participant = yield select(getParticipant, payload.participantId);

  if (participant) {
    const streams = participant.streams;

    if (streams && streams.length > 0) {
      const stream = yield call(
        getEntity,
        state,
        "streams",
        participant.streams[0]
      );

      //2- check the stream has video track
      if (stream) {
        const video = yield call(hasVideo, [stream]);

        //3- if so add stream id to stream ids array to be displayed (update conference entities)
        if (video) {
          yield put(
            addStreamToSecondStage({
              id: conferenceID,
              entityType: "conferences",
              data: {
                secondStageStreams: [participant.streams[0]],
              },
            })
          );
        }
      }
    }
  }
}

export function* stopBroadcastVideoOnSecondStage({ payload }) {
  const conferenceID = yield select(getCurrentConferenceID);

  yield put(
    removeStreamToSecondStage({
      id: conferenceID,
      entityType: "conferences",
      data: {
        secondStageStreams: [],
      },
    })
  );
}

export function* create({ payload }) {
  // get local participant
  const { participantID, entities: participantEntities } = yield call(
    getLocalParticipant
  );

  yield put(
    addParticipant({
      id: participantID,
      data: participantEntities.participants,
      entityType: "participants",
    })
  );

  yield put(application.setLocalParticipantID({ id: participantID }));

  const { entities: conferenceEntities, conferenceID } = yield call(
    createConference,
    payload
  );

  yield put(
    addConference({
      id: conferenceID,
      data: conferenceEntities.conferences,
      entityType: "conferences",
    })
  );

  yield put(application.setCurrentConferenceID({ id: conferenceID }));

  yield put(
    updateConference({
      id: conferenceID,
      entityType: "conferences",
      data: { participants: [participantID] },
    })
  );

  const constraints = {
    audio: true,
    video: false,
  };
  yield call(join, { payload: constraints });
}

export function* join({ payload }) {
  const state = yield select();
  const currentConferenceID = yield select(getCurrentConferenceID);
  const conference = yield call(
    getEntity,
    state,
    "conferences",
    currentConferenceID
  );

  const options = {
    constraints: payload,
    spatialAudio: true,
    dvwc: true,
  };

  // FIXME: do not want conference object to be modify by sdk.
  const copy = { ...conference };

  const { entities, conferenceID } = yield call(joinConference, {
    conference: copy,
    options,
  });

  //TODO: update conference, add spatial audio style status

  yield put(
    updateConference({
      id: conferenceID,
      entityType: "conferences",
      data: { params: entities.conferences[conferenceID].params },
    })
  );

  // NOTE: do not forget to update entities: object passed as ref!
  // yield put(setEntities({ entities, origin: "join conference" }));
  yield put(push(`/conference/${conference.alias}`));
}

export function* leave() {
  const { error } = yield call(leaveConference);

  if (!error) {
    yield put(application.leaveConference());
    yield put(push("/conference"));
  }
}
