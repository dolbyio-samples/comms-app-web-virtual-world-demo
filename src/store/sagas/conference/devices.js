import { all, call, put, select } from "redux-saga/effects";

import {
  listAudioDevices,
  listVideoDevices,
  selectAudioInput,
  selectAudioOutput,
  selectVideoInput,
  listAudioInputDevices,
  listAudioOutputDevices,
  listVideoInputDevices,
  getSelectedAudioInputDevice,
  getSelectedAudioOutputDevice,
  getSelectedVideoInputDevice,
} from "../../services/devices";

import { addDevice, updateParticipant } from "../../effects/conference";
import { getLocalParticipantID } from "../../reducers/application/selectors";

import { getParticipant } from "./audio";
import { displayMessage } from "./message";

export function* getAudioInputDevice() {
  const { id, entities, error } = yield call(getSelectedAudioInputDevice);
  if (error) {
    yield call(displayMessage, {
      type: "error",
      message: "could not retrieve current audio input device",
    });
    return;
  }
  if (id && entities) {
    const participantID = yield select(getLocalParticipantID);
    const participant = yield select(getParticipant, participantID);

    if (participant) {
      yield put(
        updateParticipant({
          id: participantID,
          entityType: "participants",
          data: { selectedAudioInputDeviceId: id },
        })
      );
    }
  }
}

export function* getAudioOutputDevice() {
  const { id, entities, error } = yield call(getSelectedAudioOutputDevice);
  if (error) {
    yield call(displayMessage, {
      type: "error",
      message: "could not retrieve current audio output device",
    });
    return;
  }
  if (id && entities && !error) {
    const participantID = yield select(getLocalParticipantID);
    const participant = yield select(getParticipant, participantID);

    if (participant) {
      yield put(
        updateParticipant({
          id: participantID,
          entityType: "participants",
          data: { selectedAudioOutputDeviceId: id },
        })
      );
    }
  }
}

export function* getVideoInputDevice() {
  const { id, entities, error } = yield call(getSelectedVideoInputDevice);
  if (error) {
    yield call(displayMessage, {
      type: "error",
      message: "could not retrieve current video input device",
    });
    return;
  }
  if (id && entities) {
    const participantID = yield select(getLocalParticipantID);
    const participant = yield select(getParticipant, participantID);

    if (participant) {
      yield put(
        updateParticipant({
          id: participantID,
          entityType: "participants",
          data: { selectedVideoInputDeviceId: id },
        })
      );
    }
  }
}

export function* updateDevices(devicesList, devicesIds, deviceType) {
  yield all(
    Object.entries(devicesList).map(([id, value]) =>
      put(
        addDevice({
          id,
          data: value,
          entityType: "devices",
        })
      )
    )
  );

  const participantID = yield select(getLocalParticipantID);
  const participant = yield select(getParticipant, participantID);

  if (participant) {
    yield put(
      updateParticipant({
        id: participantID,
        entityType: "participants",
        data: { [deviceType]: devicesIds },
      })
    );
  }
}

export function* enumerateAudioInputDevices() {
  const { ids, entities, error } = yield call(listAudioInputDevices);
  if (error) {
    yield call(displayMessage, {
      type: "error",
      message: "could not list audio input devices",
    });
    return;
  }
  if (ids && entities) {
    yield call(updateDevices, entities, ids, "audioInputDevices");
    yield call(getAudioInputDevice);
  }
}

export function* enumerateAudioOutputDevices() {
  const { ids, entities, error } = yield call(listAudioOutputDevices);
  if (error) {
    yield call(displayMessage, {
      type: "error",
      message: "could not list audio output devices",
    });
    return;
  }
  if (ids && entities) {
    yield call(updateDevices, entities, ids, "audioOutputDevices");
    yield call(getAudioOutputDevice);
  }
}

export function* enumerateVideoInputDevices() {
  const { ids, entities, error } = yield call(listVideoInputDevices);
  if (error) {
    yield call(displayMessage, {
      type: "error",
      message: "could not list video input devices",
    });
    return;
  }
  if (ids && entities) {
    yield call(updateDevices, entities, ids, "videoInputDevices");
    yield call(getVideoInputDevice);
  }
}

// deprecated
export function* enumerateAudioDevices() {
  const { ids, entities, error } = yield call(listAudioDevices);
  if (error) {
    yield call(displayMessage, {
      type: "error",
      message: "could not list audio devices",
    });
  }

  if (ids && entities) {
    yield all(
      Object.entries(entities).map(([id, value]) =>
        put(
          addDevice({
            id,
            data: value,
            entityType: "devices",
          })
        )
      )
    );

    const participantID = yield select(getLocalParticipantID);
    const participant = yield select(getParticipant, participantID);

    /*    const updatedDeviceIds =
      participant.audioDevices && participant.audioDevices.length > 0
        ? [...participant.audioDevices, ...ids]
        : ids;*/

    if (participant) {
      yield put(
        updateParticipant({
          id: participantID,
          entityType: "participants",
          data: { audioDevices: ids },
        })
      );
    }
  }
}

// deprecated
export function* enumerateVideoDevices() {
  const { ids, entities, error } = yield call(listVideoDevices);

  if (error) {
    yield call(displayMessage, {
      type: "error",
      message: "could not list video devices",
    });
    return;
  }

  if (ids && entities) {
    yield all(
      Object.entries(entities).map(([id, value]) =>
        put(
          addDevice({
            id,
            data: value,
            entityType: "devices",
          })
        )
      )
    );

    const participantID = yield select(getLocalParticipantID);
    const participant = yield select(getParticipant, participantID);

    const updatedDeviceIds =
      participant.videoDevices && participant.videoDevices.length > 0
        ? [...participant.videoDevices, ...ids]
        : ids;

    if (participant) {
      yield put(
        updateParticipant({
          id: participantID,
          entityType: "participants",
          data: { videoDevices: updatedDeviceIds },
        })
      );
    }
  }
}

export function* setAudioInput({ payload }) {
  const { error } = yield call(selectAudioInput, payload);
  if (error) {
    yield call(displayMessage, {
      type: "error",
      message: "could not select audio input device",
    });
  }
}

export function* setVideoInput({ payload }) {
  const { error } = yield call(selectVideoInput, payload);
  if (error) {
    yield call(displayMessage, {
      type: "error",
      message: "could not select video input device",
    });
  }
}

export function* setAudioOutput({ payload }) {
  const { error } = yield call(selectAudioOutput, payload);
  if (error) {
    yield call(displayMessage, {
      type: "error",
      message: "sdk: failed to select audio output device",
    });
  }
}
