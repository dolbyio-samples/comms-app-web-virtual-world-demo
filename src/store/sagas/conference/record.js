import { call, select, put } from "redux-saga/effects";

import { startRecording, stopRecording } from "../../services/record";
import { getEntity } from "../../reducers/entities/selectors";
import { getCurrentConferenceID } from "../../reducers/application/selectors";
import { updateConference } from "../../effects/conference";
import { displayMessage } from "./message";

export const getConference = (state, id) => getEntity(state, "conferences", id);

export function* stopRecordingConference() {
  const { error } = yield call(stopRecording);
  if (error) {
    yield call(displayMessage, {
      type: "error",
      message: "could not stop recording event",
    });

    return;
  }

  const currentConferenceID = yield select(getCurrentConferenceID);
  const conference = yield select(getConference, currentConferenceID);

  yield put(
    updateConference({
      id: conference.id,
      entityType: "conferences",
      data: { recording: false },
    })
  );
}

export function* recordConference() {
  const { error } = yield call(startRecording);
  if (error) {
    yield call(displayMessage, {
      type: "error",
      message: "could not start recording event",
    });

    return;
  }

  const currentConferenceID = yield select(getCurrentConferenceID);
  const conference = yield select(getConference, currentConferenceID);

  yield put(
    updateConference({
      id: conference.id,
      entityType: "conferences",
      data: { recording: true },
    })
  );
}
