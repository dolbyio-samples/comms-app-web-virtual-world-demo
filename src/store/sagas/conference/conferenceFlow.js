import { takeEvery, takeLatest } from "redux-saga/effects";

import { application, authentication } from "../../effects";
import { conference, devices, spatial } from "../../actions";
import { stopVideo, startVideo } from "./video";
import { startAudio, stopAudio } from "./audio";
import { recordConference, stopRecordingConference } from "./record";
import {
  create,
  join,
  leave,
  startBroadcastVideoOnMainStage,
  stopBroadcastVideoOnMainStage,
  startBroadcastVideoOnSecondStage,
  stopBroadcastVideoOnSecondStage,
} from "./conference";
import { watchParticipant } from "./speaker";
import { startScreenShare, stopScreenShare } from "./screenShare";
import {
  enumerateVideoInputDevices,
  enumerateAudioOutputDevices,
  enumerateAudioInputDevices,
  setAudioInput,
  setAudioOutput,
  setVideoInput,
  getAudioInputDevice,
  getVideoInputDevice,
  getAudioOutputDevice,
} from "./devices";
import {
  setParticipantDirection,
  setParticipantLocation,
  setParticipantPosition,
  setSpatialEnvironment,
} from "./spatial";
import { addParticipant } from "../../effects/conference";
import initialize from "./initialize";
import { watchLocations } from "./watchLocations";

export default function* conferenceFlow() {
  yield takeEvery(authentication.setToken, initialize);
  yield takeEvery(conference.create, create);
  yield takeEvery(conference.join, join);
  yield takeEvery(conference.leave, leave);
  yield takeEvery(conference.record, recordConference);
  yield takeEvery(conference.stopRecording, stopRecordingConference);
  yield takeEvery(conference.startVideo, startVideo);
  yield takeEvery(conference.stopVideo, stopVideo);
  yield takeEvery(conference.startLocalAudio, startAudio);
  yield takeEvery(conference.stopLocalAudio, stopAudio);
  yield takeEvery(addParticipant, watchParticipant);
  yield takeEvery(conference.startScreenShare, startScreenShare);
  yield takeEvery(conference.stopScreenShare, stopScreenShare);

  yield takeEvery(devices.listAudioOutputDevices, enumerateAudioOutputDevices);
  yield takeEvery(devices.listAudioInputDevices, enumerateAudioInputDevices);
  yield takeEvery(devices.listVideoInputDevices, enumerateVideoInputDevices);

  yield takeEvery(devices.getAudioInputDevice, getAudioInputDevice);
  yield takeEvery(devices.getAudioOutputDevice, getAudioOutputDevice);
  yield takeEvery(devices.getVideoInputDevice, getVideoInputDevice);

  yield takeEvery(devices.setAudioInput, setAudioInput);
  yield takeEvery(devices.setAudioOutput, setAudioOutput);
  yield takeEvery(devices.setVideoInput, setVideoInput);

  yield takeEvery(spatial.setSpatialEnvironment, setSpatialEnvironment);
  yield takeEvery(spatial.setParticipantLocation, setParticipantLocation);
  yield takeLatest(application.setCurrentConferenceID, watchLocations);
  yield takeEvery(
    conference.startVideoOnMainStage,
    startBroadcastVideoOnMainStage
  );
  yield takeEvery(
    conference.stopVideoOnMainStage,
    stopBroadcastVideoOnMainStage
  );
  yield takeEvery(
    conference.startVideoOnSecondStage,
    startBroadcastVideoOnSecondStage
  );
  yield takeEvery(
    conference.stopVideoOnSecondStage,
    stopBroadcastVideoOnSecondStage
  );
  yield takeEvery(spatial.setParticipantPosition, setParticipantPosition);
  yield takeEvery(spatial.setParticipantDirection, setParticipantDirection);
}
