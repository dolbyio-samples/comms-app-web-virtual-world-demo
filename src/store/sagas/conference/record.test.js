import {
  getConference,
  recordConference,
  stopRecordingConference,
} from "./record";
import { startRecording, stopRecording } from "../../services/record";
import { call, put, select } from "redux-saga/effects";
import { getCurrentConferenceID } from "../../reducers/application/selectors";
import { updateConference } from "../../effects/conference";
import { vi, expect, describe, it } from "vitest";
import { displayMessage } from "./message.js";

vi.mock("@voxeet/voxeet-web-sdk", () => ({
  default: {},
}));

describe("record test suite", () => {
  it("should return a conference", () => {
    const state = {
      entities: {
        conferences: {
          1: {
            alias: "test",
          },
        },
      },
    };
    const id = "1";

    const conference = getConference(state, id);

    expect(conference).toMatchSnapshot();
  });

  it("should stop recording the conference", () => {
    const currentConferenceID = "1";
    const conference = { id: currentConferenceID };
    const generator = stopRecordingConference();
    expect(generator.next().value).toEqual(call(stopRecording));
    expect(generator.next({ response: {} }).value).toEqual(
      select(getCurrentConferenceID)
    );
    expect(generator.next(currentConferenceID).value).toEqual(
      select(getConference, currentConferenceID)
    );
    expect(generator.next(conference).value).toEqual(
      put(
        updateConference({
          id: conference.id,
          entityType: "conferences",
          data: { recording: false },
        })
      )
    );
    expect(generator.next().done).toBeTruthy();
  });

  it("should start recording the conference", () => {
    // arrange
    const currentConferenceID = "1";
    const conference = { id: currentConferenceID };

    // act
    const generator = recordConference();

    // assert
    expect(generator.next().value).toEqual(call(startRecording));
    expect(generator.next({ response: {} }).value).toEqual(
      select(getCurrentConferenceID)
    );
    expect(generator.next(currentConferenceID).value).toEqual(
      select(getConference, currentConferenceID)
    );
    expect(generator.next(conference).value).toEqual(
      put(
        updateConference({
          id: conference.id,
          entityType: "conferences",
          data: { recording: true },
        })
      )
    );
    expect(generator.next().done).toBeTruthy();
  });

  it("should fail to record a conference and dispatch an action to the store", () => {
    // arrange
    const generator = recordConference();

    // act
    expect(generator.next().value).toEqual(call(startRecording));
    expect(generator.next({ error: {} }).value).toEqual(
      call(displayMessage, {
        type: "error",
        message: "could not start recording event",
      })
    );
    expect(generator.next().done).toBeTruthy();
  });

  it("should fail to stop recording a conference and dispatch an action to the store", () => {
    // arrange
    const generator = stopRecordingConference();

    // act
    expect(generator.next().value).toEqual(call(stopRecording));
    expect(generator.next({ error: {} }).value).toEqual(
      call(displayMessage, {
        type: "error",
        message: "could not stop recording event",
      })
    );
    expect(generator.next().done).toBeTruthy();
  });
});
