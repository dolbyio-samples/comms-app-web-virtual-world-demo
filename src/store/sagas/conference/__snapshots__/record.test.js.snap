// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`record test suite > should return a conference 1`] = `
{
  "alias": "test",
}
`;
